blog/_posts/2017-02-23-scalafix-v0.3.md

```diff
@@ -0,0 +1,141 @@
+The semantic API provides operations to query information from the compiler.
+
+The first version of the scala.meta semantic API makes it possible to query for the "symbol" of a name that appears in a Scala source file.
+A symbol is a unique identifier of a definition such as a class, val, def or trait.
+For example, the symbol of `println` from the standard library is `_root_.scala.Predef.println(Ljava/lang/Object;)V`.
+
+Symbols are created by "scalahost", a compiler plugin.
+Scalahost emits symbols of a source file in the form of a "semantic database".
+The semantic database can be persisted to files on disk and loaded for later analysis.
+Semantic databases from different compilation units, potentially produced by different
+versions of the Scala compiler, can be merged.
+This opens possibilities for large-scale code analysis.
+
+The introduction of the scala.meta semantic API is a game changer for scalafix.
+The ability to query for symbols open possibilities for many scalafix rewrites.
+
+## Rewrite: meta.Tree => Seq[Patch]
+
+A scalafix `Patch` is a small operation that can produce a diff on a Scala source file.
+A patch can either be a "token patch" or a "tree patch".
+
+Token patches are low-level but give full control of how small details are handled in a source files, for example formatting and comments 
+Example token patches are `Remove(token)` and `AddLeft(token, toAdd: String)`, which removes or prepends a string to `token`, respectively.
+
+Tree patches are high-level and allow rewrite author to declaratively explain what operation to perform.
+An example tree patch is `AddGlobalImport(importer)`, which adds a new import to the top of a file if it does not exist.
+Observe that `AddGlobalImport` does not worry about token-level details such as whether the user groups imports by prefix (e.g., `import a.{b, c}`).
+
+Tree and token patches build a small algebra of operations that can be composed to build complex refactorings.
+A challenge with composing patches is to figure out what to do on conflicts.
+For example, what happens when one patch renames a token while the other patch removes the same token?
+The current strategy in scalafix is to try and resolve as many conflicts as possible on the tree patch level.
+Once we we only have token patches, it is harder to resolve conflicts.
+Unsolvable conflicts on the token level abort the refactoring.
+In the future, we hope to support more advanced conflict resolution strategies.
+
+In a nutshell, a scalafix `Rewrite` is a `scala.meta.Tree => Seq[Patch]` function.
+The tree is backed by the scala.meta semantic API, so the rewrite is able to query for compiler information such as symbols.
+To demonstrate how rewrites are implemented with scalafix v0.3, let's step through an example use-case.
+
+## Example: Xor to Either
+
+The functional programming library [cats][cats] migrated recently from its `Xor` data type to `Either` from the standard library.
```

Comment:
(just a tip, in Markdown you can write [cats][] instead of [cats][cats] for the same effect ;-))